torchtune/datasets/_chat.py

```diff
@@ -81,65 +76,23 @@ def __getitem__(self, index: int) -> Tuple[List[int], List[int]]:
         return self._prepare_sample(sample)
 
     def _prepare_sample(self, sample: Mapping[str, Any]) -> Tuple[List[int], List[int]]:
-        dialogue = self._convert_to_dialogue(sample)
-
-        prompt_tokens = []
-        label_tokens = []
-        for prompt, label in self._get_turns(dialogue):
-            formatted_prompt = self.template.format(prompt)
-            encoded_prompt_with_response, labels = tokenize_prompt_and_response(
-                tokenizer=self._tokenizer,
-                prompt=formatted_prompt,
-                response=label,
-                train_on_input=self.train_on_input,
-            )
-            prompt_tokens.extend(encoded_prompt_with_response)
-            label_tokens.extend(labels)
-
-            if len(prompt_tokens) >= self.max_seq_len:
-                break
-
-        prompt_tokens, label_tokens = truncate(
-            self._tokenizer, prompt_tokens, label_tokens, self.max_seq_len
+        messages = self._convert_to_messages(sample)
+        messages = self.chat_format.format(messages)
+        tokens, mask = self._tokenizer.tokenize_messages(
```

Comment:
Where is train_on_input used for ChatDataset?

Reply:
Good point. Gonna pass it to _convert_to_messages, I think it makes most sense there tbh. Basically we want it wherever we are building the messages

Reply:
mm `_convert_to_messages` is provided by the user right? so then we have to enforce that their Callables take in this parameter and the burden is on them to mask appropriately. maybe pass it to chat_format or tokenize_messages and let the user know that if they set it it will overwrite whatever custom masking they set up in `_convert_to_messages`?

Reply:
Yeah I agree. Similar to my comment above, can we handle this in a follow-up? I think it's a relatively standalone change not too related to all this tokenizer business